File: code/data_structures/src/maxSubArray(KadaneAlgorithm)/KadaneAlgorithm.cpp
#include <bits/stdc++.h>

using namespace std;

int Kadane(vector<int>& arr) {
    int max_so_far = INT_MIN;
    int max_ending_here = 0;
    for (int i = 0; i < arr.size(); i++) {
        max_ending_here = max_ending_here + arr[i];
        if (max_ending_here < 0) {
            max_ending_here = 0;
        }
        if (max_so_far < max_ending_here) {
            max_so_far = max_ending_here;
        }
    }

    return max_so_far;
}

int main(){
    vector<int> a = {-2, -3, 4, -1, -2, 1, 5, -3};
    cout << Kadane(a) << endl;
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)

// Kadane algorithm is a simple algorithm for finding the maximum subarray sum in a given array.

File: code/data_structures/src/maxSubArray(KadaneAlgorithm)/Kadane_algorithm.py

def maxSubArraySum(a,size):
      
    max_so_far = a[0] - 1
    max_ending_here = 0
      
    for i in range(0, size):
        max_ending_here = max_ending_here + a[i]
        if (max_so_far < max_ending_here):
            max_so_far = max_ending_here
 
        if max_ending_here < 0:
            max_ending_here = 0  
    return max_so_far
  
#a = [-13, -3, -25, -20, -3, -16, -23, -12, -5, -22, -15, -4, -7]
a = [-2, -3, 4, -1, -2, 1, 5, -3]
print("Maximum contiguous sum is", maxSubArraySum(a,len(a)))
